### DISCUSSION TITLE
Combobox matches with complex list

### DISCUSSION THREAD

**Discussion by cloud-walker** [ariakit/ariakit#1868] [Questions] [answered]
Hi @diegohaz I'm struggling to understand how to implement a Combobox with matches (and so a proper searchable value), with complex lists so I can retrieve the item id later on ComboboxItem click.

An example (pseudo code):

```tsx
import {
  Combobox,
  ComboboxItem,
  ComboboxPopover,
  useComboboxState,
} from 'ariakit/combobox'
 import './style.css'

export default function App() {
  const combobox = useComboboxState({
    gutter: 8,
    sameWidth: true,
    list: [{id: '1', label: 'Searchable text'}], // here I want to pass a complex list with all useful properties for later usage
  })
  return (
    <div>
      <label className="label">
        Your favorite food
        <Combobox
          state={combobox}
          placeholder="e.g., Apple"
          className="combobox"
        />
      </label>
      <ComboboxPopover state={combobox} className="popover">
        {combobox.matches.length ? (
          combobox.matches.map((value) => (
            // here ideally i want to configure the ComboboxItem with: the id to pass to the selection event, and maybe custom content (jsx?), but still let the combobox items being searchable
            <ComboboxItem key={value} value={value} className="combobox-item" />
          ))
        ) : (
          <div className="no-results">No results found</div>
        )}
      </ComboboxPopover>
    </div>
  )
}

```

Thanks in advance!

**Comment by diegohaz** [comment, marked as answer]
`list`/`matches` is just for simple use cases, and we'll probably remove it in future versions. I recommend writing your own filter logic using something like [`match-sorter`](https://www.npmjs.com/package/match-sorter).

Example: https://codesandbox.io/s/ariakit-combobox-with-match-sorter-41cj0s?file=/src/combobox-matches.tsx:325-443

**Reply by cloud-walker** [reply]
Thanks for the response! Seems go to me, I've planned to implement a similar combobox in the next days, so I can verify the solution you propose.

**Reply by diegohaz** [reply]
I've updated the solution to `useMemo` instead of `useState` + `useEffect`.

**Reply by cloud-walker** [reply]
yeah today I've tried it, and used `useMemo` too 😄
